lib/redis_memo/memoize_query/cached_select/bind_params.rb

```diff
@@ -2,47 +2,217 @@
+    #
+    class Plan
+      class DependencySizeEstimation
+        def initialize(hash = nil)
+          @hash = hash
+        end
+
+        def +(other)
+          merged_hash = hash.dup
+          other.hash.each do |k, v|
+            merged_hash[k] += v
+          end
+          self.class.new(merged_hash)
+        end
+
+        def *(other)
+          merged_hash = hash.dup
+          other.hash.each do |k, v|
+            if merged_hash.include?(k)
+              merged_hash[k] *= v
+            else
+              merged_hash[k] = v
+            end
+          end
+          self.class.new(merged_hash)
+        end
+
+        def [](key)
+          hash[key]
+        end
+
+        def []=(key, val)
+          hash[key] = val
+        end
+
+        def to_i
+          ret = 0
+          hash.each do |_, v|
+            ret += v
+          end
+          ret
+        end
 
-      # The tree is almost always right-heavy. Merge into the right node for better
-      # performance.
-      other.params.merge!(params) do |_, other_attrs_set, attrs_set|
-        if other_attrs_set.empty?
-          attrs_set
-        elsif attrs_set.empty?
-          other_attrs_set
-        else
-          attrs_set + other_attrs_set
+        protected
+
+        def hash
+          @hash ||= Hash.new(0)
         end
       end
 
-      other
+      attr_accessor :dependency_size_estimation
+      attr_accessor :model_attrs
+
+      def initialize(bind_params)
+        @dependency_size_estimation = DependencySizeEstimation.new
+        @model_attrs = Hash.new do |models, model|
+          models[model] = Set.new
+        end
+
+        # An aggregated bind_params node can only obtain params by combining
+        # its children nodes
+        return if !bind_params.__send__(:operator).nil?
+
+        bind_params.params.each do |model, attrs_set|
+          @dependency_size_estimation[model] += attrs_set.size
+          attrs_set.each do |attrs|
+            # [k, nil]: Ignore the attr value and keep the name only
+            @model_attrs[model] << attrs.keys.map { |k| [k, nil] }.to_h
+          end
+        end
+      end
     end
 
-    def product(other)
+    attr_accessor :plan
+
+    def plan!
+      self.plan = Plan.new(self)
+      return if operator.nil?
+
+      left.plan!
+      right.plan!
+      __send__(:"plan_#{operator}")
+    end
+
+    def plan_union
+      plan.dependency_size_estimation = left.plan.dependency_size_estimation + right.plan.dependency_size_estimation
+      plan.model_attrs = union_attrs_set(left.plan.model_attrs, right.plan.model_attrs)
+    end
+
+    def plan_product
+      plan.dependency_size_estimation = left.plan.dependency_size_estimation * right.plan.dependency_size_estimation
+      plan.model_attrs = product_attrs_set(left.plan.model_attrs, right.plan.model_attrs)
+    end
+
+    def union!
+      @params = union_attrs_set(left.params, right.params)
+    end
```

Comment:
if this union is now used in the extraction, can we name it as `extrsction_union`?

Reply:
I think the current name makes a lot more sense to me. As in we are planning for an action, before actually perform an action